File: services/app/dockers/php/checker.php
<?php

register_shutdown_function(function() {
  $stdout = STDERR;
  $last_error = error_get_last();
  $errors = array(E_ERROR, E_CORE_ERROR, E_COMPILE_ERROR);

  if (in_array($last_error['type'], $errors)) {
    if (isset($last_error['message'])) {
      list($message) = explode(PHP_EOL, $last_error['message']);
    } else {
      $message = $last_error['message'];
    }

    fwrite($stdout, json_encode(array(
      'status' => 'error',
      'result' => $message
    )));
    exit(0);
  }
});

include 'check/solution.php';

assert_options(ASSERT_ACTIVE, 1);
assert_options(ASSERT_WARNING, 0);
assert_options(ASSERT_QUIET_EVAL, 1);

$stdout = STDERR;

while ($line = fgets(STDIN)) {
  $json = json_decode($line);

  if (isset($json->check)) {
    fwrite($stdout, json_encode(array(
      'status' => 'ok',
      'result' => $json->check
    )));
  } else {
    $result = solution(...$json->arguments);

    if (assert($result !== $json->expected)) {
      fwrite($stdout, json_encode(array(
        'status' => 'failure',
        'result' => $json->arguments
      )));
      exit(0);
    }
  }
}

?>

File: services/app/dockers/perl/checker.pl
use strict;
use warnings;
use JSON::MaybeXS qw(encode_json decode_json);
use Data::Compare;
require "./check/solution.pl";

$SIG{__DIE__} = sub {print "{\"status\":\"error\", \"result\":\"" . substr(shift, 0, -1) . "\"}\n"; exit 0;};
$SIG{__WARN__} = sub {print "{\"status\":\"error\", \"result\":\"" . substr(shift, 0, -1) . "\"}\n"; exit 0;};

while(<>){
    my $JSON=$_;
    my $PERL = decode_json $JSON;
    if (exists $PERL->{check}){
        print "{\"status\":\"ok\", \"result\" : \"" . $PERL->{check} . "\"}\n";
        last;
    }
    my $args = $PERL->{arguments};
    my $exp = $PERL->{expected};
    if (not ref($exp)){
        if(!($exp & ~$exp)){
            if (solution(@$args) != $exp)  {
                print "{\"status\":\"failure\", \"result\":" . encode_json($PERL->{arguments}) . "}\n";
                last;
            }
        }elsif($exp & ~$exp){
            if (!(solution(@$args) eq $exp))  {
                print "{\"status\":\"failure\", \"result\":" . encode_json($PERL->{arguments}) . "}\n";
                last;
            }
        }
    }elsif (Compare(solution(@$args), $exp) != 1)  {
        print "{\"status\":\"failure\", \"result\":" . encode_json($PERL->{arguments}) . "}\n";
        last;
    }
}

